File: 138. Copy List with Random Pointer/Java/Solution.java
/**
 * Definition for singly-linked list with a random pointer.
 * class RandomListNode {
 *     int label;
 *     RandomListNode next, random;
 *     RandomListNode(int x) { this.label = x; }
 * };
 */
public class Solution {
    public RandomListNode copyRandomList(RandomListNode head) {
        if (head == null) {
            return head;
        }
        Map<RandomListNode, RandomListNode> memory = new HashMap<>();
        RandomListNode newHead = new RandomListNode(head.label);
        RandomListNode current = head;
        RandomListNode newCurrent = newHead;
        memory.put(current, newCurrent);
        while (current.next != null) {
            newCurrent.next = new RandomListNode(current.next.label);
            memory.put(current.next, newCurrent.next);
            newCurrent = newCurrent.next;
            current = current.next;
        }
        current = head;
        newCurrent = newHead;
        while (current != null) {
            if (current.random != null) {
                newCurrent.random = memory.get(current.random);
            }
            newCurrent = newCurrent.next;
            current = current.next;
        }
        return newHead;
    }
}

File: 138. Copy List with Random Pointer/Python/Solution.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None


class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
        if not head:
            return head
        memory = {}
        new_head = RandomListNode(head.label)
        current = head
        new_current = new_head
        memory[current] = new_current
        while current.next:
            new_current.next = RandomListNode(current.next.label)
            memory[current.next] = new_current.next
            new_current = new_current.next
            current = current.next
        current = head
        new_current = new_head
        while current:
            if current.random:
                new_current.random = memory[current.random]
            new_current = new_current.next
            current = current.next
        return new_head
